### PR TITLE
plan, executor, expression: refactor explain

### PR DESCRIPTION
**Pull request by zz-jason**

### CODE REVIEW

**Comment by winoros** [comment]
What the relation is between this pr and https://github.com/pingcap/tidb/pull/3625 ?

**Comment by zz-jason** [comment]
https://github.com/pingcap/tidb/pull/3625 should be abandoned. `MarshalJSON()` is reserved to print more detailed informations about physical plan in log.

This is a draft for new explain framework, we need to implement `ExplainInfo` for operators, expressions, aggregations(and window functions in the future). If this framework is ok, I'll fix unit test and mysql test, submit this change, and implement `ExplainInfo()` for every operator.

@hanfei1991 PTAL

File: executor/explain.go [hunk 1]
```diff
@@ -59,12 +61,80 @@ func (e *ExplainExec) prepareExplainInfo(p plan.Plan, parent plan.Plan) error {
...
+// prepareExplainInfo4DAGTask generates the following informations for every plan:
+// ["id", "parents", "schema", "key", "task", "operator info"]
+func (e *ExplainExec) prepareExplainInfo4DAGTask(p plan.PhysicalPlan, taskType string) error {
+	parents := p.Parents()
+	parentIDs := make([]string, 0, len(parents))
+	for _, parent := range parents {
+		parentIDs = append(parentIDs, parent.ID())
+	}
+	parentInfo := strings.Join(parentIDs, ",")
+	columnInfo := p.Schema().ColumnInfo()
```
Comment (hanfei1991) [inline executor/explain.go]: I prefer not to print shema and key info.

File: executor/explain.go [hunk 2]
```diff
@@ -59,12 +61,80 @@ func (e *ExplainExec) prepareExplainInfo(p plan.Plan, parent plan.Plan) error {
...
 // Next implements Execution Next interface.
 func (e *ExplainExec) Next() (*Row, error) {
 	if e.cursor == 0 {
-		err := e.prepareExplainInfo(e.StmtPlan, nil)
-		if err != nil {
-			return nil, errors.Trace(err)
+		if plan.UseDAGPlanBuilder(e.ctx) {
+			e.StmtPlan.SetParents()
+			e.explainedPlans = map[string]bool{}
+			err := e.prepareRootTaskInfo(e.StmtPlan.(plan.PhysicalPlan))
```
Comment (hanfei1991) [inline executor/explain.go]: Can we do preparation in planner ?

Comment (zz-jason) [inline executor/explain.go]: It's only used by `ExplainExec`, why should we move it to planner ?

Comment (hanfei1991) [inline executor/explain.go]: ExplainExec only genarated by Explain plan.

Comment (zz-jason) [inline executor/explain.go]: moved

File: executor/explain.go (hunk 1 above)
Comment (zz-jason) [inline executor/explain.go]: removed

**Comment by zz-jason** [comment]
@hanfei1991 PTAL

File: plan/optimizer.go [hunk 3]
```diff
@@ -113,6 +113,43 @@ func checkPrivilege(pm privilege.Manager, vs []visitInfo) bool {
 	return true
 }
 
+func setParents4FinalPlan(plan PhysicalPlan) {
```
Comment (hanfei1991) [inline plan/optimizer.go]: It's better not to put it in this file. This is only for explain. We can rename `explain_info.go` to `explain.go` and remove this funtion to that file.

Comment (zz-jason) [inline plan/optimizer.go]: addressed

**Comment by hanfei1991** [comment]
LGTM

**Comment by shenli** [comment]
Should we update tidb-test?
@winoros PTAL

**Comment by zz-jason** [comment]
no need to, not affected

File: expression/explain.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	"bytes"
+	"fmt"
+)
+
+// ExplainInfo implements the Expression interface.
+func (expr *ScalarFunction) ExplainInfo() string {
+	buffer := bytes.NewBufferString(fmt.Sprintf("%s(", expr.FuncName.L))
+	for i, arg := range expr.GetArgs() {
+		buffer.WriteString(fmt.Sprintf("%s", arg.ExplainInfo()))
+		if i < len(expr.GetArgs()) {
```
Comment (lamxTyler) [inline expression/explain.go]: This should be `i+1<...`?

File: expression/expression.go [hunk 5]
```diff
@@ -98,6 +98,9 @@ type Expression interface {
 
 	// ResolveIndices resolves indices by the given schema.
 	ResolveIndices(schema *Schema)
+
+	// ExplainInfo returns operator informations to be explained
```
Comment (lamxTyler) [inline expression/expression.go]: add `.`.

File: plan/plan.go [hunk 6]
```diff
@@ -264,6 +264,9 @@ type PhysicalPlan interface {
 	// ToPB converts physical plan to tipb executor.
 	ToPB(ctx context.Context) (*tipb.Executor, error)
 
+	// ExplainInfo returns operator informations to be explained
```
Comment (lamxTyler) [inline plan/plan.go]: add `.`

File: plan/plans.go [hunk 7]
```diff
@@ -160,5 +165,70 @@ type DDL struct {
...
+	if parent != nil {
+		parentStr = parent.ID()
+	}
+	row := types.MakeDatums(p.ID(), string(explain), parentStr)
+	e.Rows = append(e.Rows, row)
+	return nil
+}
+
+// prepareExplainInfo4DAGTask generates the following informations for every plan:
+// ["id", "parents", "schema", "key", "task", "operator info"]
```
Comment (lamxTyler) [inline plan/plans.go]: add `.`

File: plan/explain.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+// limitations under the License.
+
+package plan
+
+import (
+	"fmt"
+)
+
+func setParents4FinalPlan(plan PhysicalPlan) {
+	allPlans := []PhysicalPlan{}
```
Comment (winoros) [inline plan/explain.go]: just initialize the `allPlans` with `plan`, no need to append?

File: plan/plan.go (hunk 6 above)
Comment (winoros) [inline plan/plan.go]: `information` should be a uncountable noun?

File: expression/explain.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+// ExplainInfo implements the Expression interface.
+func (expr *Constant) ExplainInfo() string {
+	valStr, err := expr.Value.ToString()
+	if err != nil {
+		valStr = "not recognized const value"
+	}
+	return fmt.Sprintf("const(%s)", valStr)
+}
+
+// ExplainAggFunc generates explain informations for a aggregation function.
```
Comment (winoros) [inline expression/explain.go]: `information` should be an uncountable noun?

Comment (zz-jason) [inline expression/explain.go]: done

File: plan/explain.go (hunk 8 above)
Comment (zz-jason) [inline plan/explain.go]: done

**Comment by zz-jason** [comment]
@winoros @lamxTyler PTAL

**Comment by zz-jason** [comment]
to https://github.com/pingcap/tidb/issues/3805

**Pull request merged**